docs/End-To-End-Encryption.md

```diff
@@ -0,0 +1,181 @@
+
+PowerAuth supports a standard ECIES encryption (integrated encryption scheme that uses elliptic curve cryptography) with the standard X9.63 (SHA256) KDF function (that produces 32b long keys).
+
+### ECIES Encryption
+
+Assume we have a public key `KEY_ENC_PUB`, data `DATA_ORIG` to be encrypted and a `SHARED_INFO_1` and `SHARED_INFO_2` constants (`byte[]`) as an encryption parameters. ECIES encryption works in a following way:
+
+1. Generate an ephemeral key pair:
+    ```java
+    EPH_KEYPAIR = (KEY_EPH_PRIV, KEY_EPH_PUB).
+    ```
+2. Derive base secret key (in this step, we do not trim the key to 16b only, we keep all 32b).
+    ```java
+    SecretKey KEY_BASE = ECDH.phase(KEY_EPH_PRIV, KEY_ENC_PUB)
+    ```
+3. Derive a secret key using X9.63 KDF function (using SHA256 internally). When calling the KDF, we use `SHARED_INFO_1` together with `KEY_EPH_PUB` value (as raw `byte[]`) as an `info` parameter.
+    ```java
+    byte[] INFO = Bytes.concat(SHARED_INFO_1, KEY_EPH_PUB);
+    SecretKey KEY_SECRET = KDF_X9_63_SHA256.derive(KEY_BASE, INFO)
+    ```
+4. Split the 32b long `KEY_SECRET` to two 16b keys. The first part is used as an encryption key `KEY_ENC`. The second part is used as MAC key `KEY_MAC`.
+    ```java
+    SecretKey KEY_ENC = ByteUtils.subarray(KEY_SECRET, 0, 16);
+    SecretKey KEY_MAC = ByteUtils.subarray(KEY_SECRET, 16, 16);
+    ```
+5. Compute the encrypted data using AES, with zero `iv` value.
+    ```java
+    byte[] iv = ByteUtils.zeroBytes(16);
+    byte[] DATA_ENCRYPTED = AES.encrypt(DATA_ORIG, iv, KEY_ENC)
+    ```
+6. Compute the MAC of encrypted data, include `SHARED_INFO_2`.
+    ```java
+    byte[] DATA = Bytes.concat(DATA_ENCRYPTED, SHARED_INFO_2);
+    byte[] MAC = Mac.hmacSha256(KEY_MAC, DATA)
+    ```
+7. Prepare ECIES payload.
+    ```java
+    EciesPayload payload = (DATA_ENCRYPTED, MAC, KEY_EPH_PUB)
+    ```
+
+### ECIES Decryption
+
+Assume we have a private key `KEY_ENC_PRIV`, encrypted data as an instance of the ECIES payload `(DATA_ENCRYPTED, MAC, KEY_EPH_PUB)` and a `SHARED_INFO_1` and `SHARED_INFO_2` constants (`byte[]`) as an decryption parameters. ECIES decryption works in a following way:
+
+1. Derive base secret key from the private key and ephemeral public key from the ECIES payload (in this step, we do not trim the key to 16b only, we keep all 32b).
+    ```java
+    SecretKey KEY_BASE = ECDH.phase(KEY_ENC_PRIV, KEY_EPH_PUB)
+    ```
+2. Derive a secret key using X9.63 KDF function (using SHA256 internally). When calling the KDF, we use `KEY_EPH_PUB` value (as raw `byte[]`) as an `info` parameter.
+    ```java
+    byte[] INFO = Bytes.concat(SHARED_INFO_1, KEY_EPH_PUB);
+    SecretKey KEY_SECRET = KDF_X9_63_SHA256.derive(KEY_BASE, INFO)
+    ```
+3. Split the 32b long `KEY_SECRET` to two 16b keys. The first part is used as an encryption key `KEY_ENC`. The second part is used as MAC key `KEY_MAC`.
+    ```java
+    SecretKey KEY_ENC = ByteUtils.subarray(KEY_SECRET, 0, 16);
```

Comment:
Add  `keyConverter.convertBytesToSharedSecretKey()`

Reply:
The same as above. It's specific for Java implementaiton.